File: port_java_OBSOLUTE/com/boare/vsq/VsqCommon.java
/*
 * VsqCommon.java
 * Copyright (c) 2008-2009 kbinani
 *
 * This file is part of com.boare.vsq.
 *
 * com.boare.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * com.boare.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
package com.boare.vsq;

import java.awt.*;

/**
 * vsqファイルのメタテキストの[Common]セクションに記録される内容を取り扱う
 */
public class VsqCommon implements Cloneable {
    public String version;
    public String name;
    public String color;
    public int dynamicsMode;
    public int playMode;

    public Object clone() {
        String[] spl = color.split( ",", 3 );
        int r = Integer.parseInt( spl[0] );
        int g = Integer.parseInt( spl[1] );
        int b = Integer.parseInt( spl[2] );
        Color color_ = new Color( r, g, b );
        VsqCommon res = new VsqCommon( name, color_, dynamicsMode, playMode );
        res.version = version;
        return res;
    }

    public static String getXmlElementName( String name ){
        if( name.equals( "version" ) ){
            return "Version";
        }else if( name.equals( "name" ) ){
            return "Name";
        }else if( name.equals( "color" ) ){
            return "Color";
        }else if( name.equals( "dynamicsMode" ) ){
            return "DynamicsMode";
        }else if( name.equals( "playMode" ) ){
            return "PlayMode";
        }
        return name;
    }

    /// <summary>
    /// 各パラメータを指定したコンストラクタ
    /// </summary>
    /// <param name="name">トラック名</param>
    /// <param name="color">Color値（意味は不明）</param>
    /// <param name="dynamics_mode">DynamicsMode（デフォルトは1）</param>
    /// <param name="play_mode">PlayMode（デフォルトは1）</param>
    public VsqCommon( String name_, Color color_, int dynamics_mode, int play_mode ) {
        this.version = "DSB301";
        this.name = name;
        this.color = color_.getRed() + "," + color_.getGreen() + "," + color_.getBlue();
        this.dynamicsMode = dynamics_mode;
        this.playMode = play_mode;
    }


    /// <summary>
    /// MetaTextのテキストファイルからのコンストラクタ
    /// </summary>
    /// <param name="sr">読み込むテキストファイル</param>
    /// <param name="last_line">読み込んだ最後の行が返される</param>
    public VsqCommon( TextMemoryStream sr, StringBuilder last_line ) {
        version = "";
        name = "";
        color = "0,0,0";
        dynamicsMode = 0;
        playMode = 0;
        String line;
        String[] spl;
        while( (line = sr.readLine()) != null ){
            if( line.startsWith( "[" ) ){
                break;
            }
            spl = last_line.toString().split( "=" );
            if( spl[0].equals( "Version" ) ){
                this.version = spl[1];
            }else if( spl[0].equals( "Name" ) ){
                this.name = spl[1];
            }else if( spl[0].equals( "Color" ) ){
                this.color = spl[1];
            }else if( spl[0].equals( "DynamicsMode" ) ){
                this.dynamicsMode = Integer.parseInt( spl[1] );
            }else if( spl[0].equals( "PlayMode" ) ){
                this.playMode = Integer.parseInt( spl[1] );
            }
        }
        last_line.setLength( 0 );
        last_line.append( line );
    }


    /// <summary>
    /// インスタンスの内容をテキストファイルに出力します
    /// </summary>
    /// <param name="sw">出力先</param>
    public void write( TextMemoryStream sw ) {
        sw.writeLine( "[Common]" );
        sw.writeLine( "Version=" + version );
        sw.writeLine( "Name=" + name );
        sw.writeLine( "Color=" + color );
        sw.writeLine( "DynamicsMode=" + dynamicsMode );
        sw.writeLine( "PlayMode=" + playMode );
    }
}

File: src/cadencii.vsq/VsqCommon.cs
/*
 * VsqCommon.cs
 * Copyright © 2008-2011 kbinani
 *
 * This file is part of cadencii.vsq.
 *
 * cadencii.vsq is free software; you can redistribute it and/or
 * modify it under the terms of the BSD License.
 *
 * cadencii.vsq is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
*/
using System;
using cadencii;
using cadencii.java.io;

namespace cadencii.vsq
{

    /// <summary>
    /// vsqファイルのメタテキストの[Common]セクションに記録される内容を取り扱う
    /// </summary>
    [Serializable]
    public class VsqCommon : ICloneable
    {
        public string Version;
        public string Name;
        public string Color;
        /// <summary>
        /// Dynamicsカーブを表示するモード(Expert)なら1、しない(Standard)なら0。
        /// </summary>
        public int DynamicsMode = cadencii.vsq.DynamicsMode.Expert;
        /// <summary>
        /// Play With Synthesisなら1、Play After Synthesiなら0、Offなら-1。
        /// </summary>
        public int PlayMode = cadencii.vsq.PlayMode.PlayWithSynth;
        /// <summary>
        /// PlayModeがOff(-1)にされる直前に，PlayAfterSynthかPlayWithSynthのどちらが指定されていたかを記憶しておく．
        /// </summary>
        public int LastPlayMode = cadencii.vsq.PlayMode.PlayWithSynth;

        public object Clone()
        {
            return clone();
        }

        public Object clone()
        {
            string[] spl = PortUtil.splitString(Color, new char[] { ',' }, 3);
            int r = int.Parse(spl[0]);
            int g = int.Parse(spl[1]);
            int b = int.Parse(spl[2]);
            VsqCommon res = new VsqCommon(Name, r, g, b, DynamicsMode, PlayMode);
            res.Version = Version;
            res.LastPlayMode = LastPlayMode;
            return res;
        }

        /// <summary>
        /// 各パラメータを指定したコンストラクタ
        /// </summary>
        /// <param name="name">トラック名</param>
        /// <param name="color">Color値（意味は不明）</param>
        /// <param name="dynamics_mode">DynamicsMode（デフォルトは1）</param>
        /// <param name="play_mode">PlayMode（デフォルトは1）</param>
        public VsqCommon(string name, int red, int green, int blue, int dynamics_mode, int play_mode)
        {
            this.Version = "DSB301";
            this.Name = name;
            this.Color = red + "," + green + "," + blue;
            this.DynamicsMode = dynamics_mode;
            this.PlayMode = play_mode;
        }

        public VsqCommon()
            : this("Miku", 179, 181, 123, 1, 1)
        {
        }

        /// <summary>
        /// MetaTextのテキストファイルからのコンストラクタ
        /// </summary>
        /// <param name="sr">読み込むテキストファイル</param>
        /// <param name="last_line">読み込んだ最後の行が返される</param>
        public VsqCommon(TextStream sr, ByRef<string> last_line)
        {
            Version = "";
            Name = "";
            Color = "0,0,0";
            DynamicsMode = 0;
            PlayMode = 1;
            last_line.value = sr.readLine();
            string[] spl;
            while (!last_line.value.StartsWith("[")) {
                spl = PortUtil.splitString(last_line.value, new char[] { '=' });
                string search = spl[0];
                if (search.Equals("Version")) {
                    this.Version = spl[1];
                } else if (search.Equals("Name")) {
                    this.Name = spl[1];
                } else if (search.Equals("Color")) {
                    this.Color = spl[1];
                } else if (search.Equals("DynamicsMode")) {
                    this.DynamicsMode = int.Parse(spl[1]);
                } else if (search.Equals("PlayMode")) {
                    this.PlayMode = int.Parse(spl[1]);
                }
                if (!sr.ready()) {
                    break;
                }
                last_line.value = sr.readLine();
            }
        }

        /// <summary>
        /// インスタンスの内容をテキストファイルに出力します
        /// </summary>
        /// <param name="sw">出力先</param>
        public void write(ITextWriter sw)
        {
            sw.writeLine("[Common]");
            sw.writeLine("Version=" + Version);
            sw.writeLine("Name=" + Name);
            sw.writeLine("Color=" + Color);
            sw.writeLine("DynamicsMode=" + DynamicsMode);
            sw.writeLine("PlayMode=" + PlayMode);
        }
    }

}
